File: generated-code/codecraft/rust/src/model/entity_properties.rs
use super::*;

#[derive(Clone, Debug)]
pub struct EntityProperties {
    pub size: i32,
    pub build_score: i32,
    pub destroy_score: i32,
    pub can_move: bool,
    pub population_provide: i32,
    pub population_use: i32,
    pub max_health: i32,
    pub initial_cost: i32,
    pub sight_range: i32,
    pub resource_per_health: i32,
    pub build: Option<BuildProperties>,
    pub attack: Option<AttackProperties>,
    pub repair: Option<RepairProperties>,
}

impl trans::Trans for EntityProperties {
    fn write_to(&self, writer: &mut dyn std::io::Write) -> std::io::Result<()> {
        self.size.write_to(writer)?;
        self.build_score.write_to(writer)?;
        self.destroy_score.write_to(writer)?;
        self.can_move.write_to(writer)?;
        self.population_provide.write_to(writer)?;
        self.population_use.write_to(writer)?;
        self.max_health.write_to(writer)?;
        self.initial_cost.write_to(writer)?;
        self.sight_range.write_to(writer)?;
        self.resource_per_health.write_to(writer)?;
        self.build.write_to(writer)?;
        self.attack.write_to(writer)?;
        self.repair.write_to(writer)?;
        Ok(())
    }
    fn read_from(reader: &mut dyn std::io::Read) -> std::io::Result<Self> {
        let size: i32 = trans::Trans::read_from(reader)?;
        let build_score: i32 = trans::Trans::read_from(reader)?;
        let destroy_score: i32 = trans::Trans::read_from(reader)?;
        let can_move: bool = trans::Trans::read_from(reader)?;
        let population_provide: i32 = trans::Trans::read_from(reader)?;
        let population_use: i32 = trans::Trans::read_from(reader)?;
        let max_health: i32 = trans::Trans::read_from(reader)?;
        let initial_cost: i32 = trans::Trans::read_from(reader)?;
        let sight_range: i32 = trans::Trans::read_from(reader)?;
        let resource_per_health: i32 = trans::Trans::read_from(reader)?;
        let build: Option<BuildProperties> = trans::Trans::read_from(reader)?;
        let attack: Option<AttackProperties> = trans::Trans::read_from(reader)?;
        let repair: Option<RepairProperties> = trans::Trans::read_from(reader)?;
        Ok(Self {
            size,
            build_score,
            destroy_score,
            can_move,
            population_provide,
            population_use,
            max_health,
            initial_cost,
            sight_range,
            resource_per_health,
            build,
            attack,
            repair,
        })
    }
}

File: generated-code/codecraft/swift/Sources/TransGenTest/EntityProperties.swift
public struct EntityProperties {
    let size: Int32
    let buildScore: Int32
    let destroyScore: Int32
    let canMove: Bool
    let populationProvide: Int32
    let populationUse: Int32
    let maxHealth: Int32
    let initialCost: Int32
    let sightRange: Int32
    let resourcePerHealth: Int32
    let build: BuildProperties?
    let attack: AttackProperties?
    let repair: RepairProperties?

    static func readFrom<S: InputStream>(_ stream: S) -> EntityProperties {
        var size: Int32
        size = stream.readInt32()
        var buildScore: Int32
        buildScore = stream.readInt32()
        var destroyScore: Int32
        destroyScore = stream.readInt32()
        var canMove: Bool
        canMove = stream.readBool()
        var populationProvide: Int32
        populationProvide = stream.readInt32()
        var populationUse: Int32
        populationUse = stream.readInt32()
        var maxHealth: Int32
        maxHealth = stream.readInt32()
        var initialCost: Int32
        initialCost = stream.readInt32()
        var sightRange: Int32
        sightRange = stream.readInt32()
        var resourcePerHealth: Int32
        resourcePerHealth = stream.readInt32()
        var build: BuildProperties?
        if stream.readBool() {
            build = BuildProperties.readFrom(stream)
        } else {
            build = nil
        }
        var attack: AttackProperties?
        if stream.readBool() {
            attack = AttackProperties.readFrom(stream)
        } else {
            attack = nil
        }
        var repair: RepairProperties?
        if stream.readBool() {
            repair = RepairProperties.readFrom(stream)
        } else {
            repair = nil
        }
        return EntityProperties(size: size, buildScore: buildScore, destroyScore: destroyScore, canMove: canMove, populationProvide: populationProvide, populationUse: populationUse, maxHealth: maxHealth, initialCost: initialCost, sightRange: sightRange, resourcePerHealth: resourcePerHealth, build: build, attack: attack, repair: repair)
    }

    func writeTo<S: OutputStream>(_ stream: S) {
        stream.writeInt32(size)
        stream.writeInt32(buildScore)
        stream.writeInt32(destroyScore)
        stream.writeBool(canMove)
        stream.writeInt32(populationProvide)
        stream.writeInt32(populationUse)
        stream.writeInt32(maxHealth)
        stream.writeInt32(initialCost)
        stream.writeInt32(sightRange)
        stream.writeInt32(resourcePerHealth)
        if build == nil {
            stream.writeBool(false)
        } else {
            stream.writeBool(true)
            let buildValue = build!
            buildValue.writeTo(stream)
        }
        if attack == nil {
            stream.writeBool(false)
        } else {
            stream.writeBool(true)
            let attackValue = attack!
            attackValue.writeTo(stream)
        }
        if repair == nil {
            stream.writeBool(false)
        } else {
            stream.writeBool(true)
            let repairValue = repair!
            repairValue.writeTo(stream)
        }
    }
}